pkg/batchfetcher/batchfetcher.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/modules"
+	"github.com/filecoin-project/mir/pkg/pb/availabilitypb"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"github.com/filecoin-project/mir/pkg/pb/requestpb"
+	t "github.com/filecoin-project/mir/pkg/types"
+)
+
+// NewModule returns a new batch fetcher module.
+// The batch receives events output by the ordering protocol (e.g. ISS)
+// and relays them to the application in the same order.
+// It replaces the DeliverCert events from the input stream by the corresponding ProvideTransactions
+// that it obtains from the availability layer.
+// It keeps track of the current epoch (by observing the relayed NewEpoch events)
+// and automatically requests the transactions from the correct instance of the availability module.
+func NewModule(mc *ModuleConfig) modules.Module {
+	m := dsl.NewModule(mc.Self)
+
+	// The current epoch number, as announced by the ordering protocol.
+	epochNr := t.EpochNr(0)
+
+	// Map of delivered requests that is used to filter duplicates.
+	// TODO: Implement compaction (client watermarks) so that this map does not grow indefinitely.
+	delivered := make(map[t.ClientID]map[t.ReqNo]struct{})
+
+	// Queue of output events. It is required for buffering events being relayed
+	// in case a DeliverCert event that has arrived before has not yet been transformed to a ProvideTransactions event.
+	var output outputQueue
+
+	// The NewEpoch handler updates the current epoch number and forwards the event to the output.
+	dsl.UponEvent[*eventpb.Event_NewEpoch](m, func(newEpoch *eventpb.NewEpoch) error {
+		epochNr = t.EpochNr(newEpoch.EpochNr)
+		output.Enqueue(&outputItem{
+			event: events.NewEpoch(mc.Destination, t.EpochNr(newEpoch.EpochNr)),
+		})
+		output.Flush(m)
+		return nil
+	})
+
+	// The DeliverCert handler requests the transactions referenced by the received availability certificate
+	// from the availability layer.
+	// TODO: Verify Certificate? Here or elsewhere?
+	dsl.UponEvent[*eventpb.Event_DeliverCert](m, func(cert *eventpb.DeliverCert) error {
+		// Skip padding certificates. DeliverCert events with nil certificates are considered noops.
+		if cert.Cert.Type == nil {
+			return nil
+		}
+
+		switch c := cert.Cert.Type.(type) {
+		case *availabilitypb.Cert_Msc:
+
+			// TODO: Check whether this makes any sense.
+			if len(c.Msc.BatchId) == 0 {
+				fmt.Println("Received empty batch availability certificate.")
+				return nil
+			}
+
+			// Create an empty output item and enqueue it immediately.
+			// Actual output will be delayed until the transactions have been received.
+			// This is necessary to preserve the order of incoming and outgoing events.
+			item := outputItem{event: nil}
+			output.Enqueue(&item)
+
+			// Request transactions from the availability layer.
+			availabilitydsl.RequestTransactions(
+				m,
+				mc.Availability.Then(t.ModuleID(fmt.Sprintf("%v", epochNr))),
+				cert.Cert,
+				&item,
```

Comment:
I usually prefer having a separate struct for the context to be more explicit and avoid accidentally using the same context type in two places.

Reply:
Yes, I see your point. It does take away some of the beauty of how the context store is used under the hood, but making things explicit usually tends to be a good idea.
Wrapped it in a separate struct.

Reply:
Thanks! Yes, this is not ideal, I agree. I spent a good amount of time trying to figure out a nicer way to do this, but it tends to quickly get a bit too complicated internally (which is not necessarily a bad thing, but I think for now it makes sense to keep things relatively low-level).